File: 09_Generics/02_variance/java/ProdCons.java
import java.util.function.*;

interface Consumer<T> {
	void digest(T obj);
}

interface Producer<T> {
	T invent();
}

class A {}
class B extends A {}
class C extends B {}

public class ProdCons<I,O> implements Consumer<I>, Producer<O> {
	
	private final Function<? super I, ? extends O> transformer;
	
	private O obj = null;
	
	public ProdCons(Function<? super I, ? extends O> transformer) {
		this.transformer = transformer;
	}
	
	@Override
	public void digest(I iobj) {
		this.obj = transformer.apply(iobj);
	}
	
	@Override
	public O invent() {
		return this.obj;
	}
	
	public static void main(String[] args) {
		
		//final ProdCons<B,B> pc = new ProdCons<B,B>((B b) -> b);
		final ProdCons<B,B> pc =
			new ProdCons<B,B>((A a) -> new C());

		pc.digest(new B());
		
		B value = pc.invent();
		
		System.out.println(value);
	}
}

File: 09_Generics/02_variance/kotlin/ProdCons.kt
interface Consumer<in T> {
    fun digest(iobj: T)
}

interface Producer<out T> {
    fun invent(): T
}

open class A
open class B : A()
open class C : B()

class ProdCons<in I : Any, out O : Any>(
		private val transformer: (i : I) -> O
	) : Consumer<I>, Producer<O> {

    private lateinit var obj: O

    override fun digest(iobj: I) {
        obj = transformer(iobj)
    }

    override fun invent(): O {
        return obj
    }

}

fun main() {
    //val pc = ProdCons<B,B> { b : B -> b }
    val pc : ProdCons<C,A> = ProdCons<A,C> { _ : A -> C() }
    pc.digest(C())
    val value : A = pc.invent()
    println(value)
}
